### ISSUE TITLE
Is it possible to get the time component with an all day DTStart and Until ?

### ISSUE THREAD

**Issue by chungonn** [dmfs/lib-recur#109]
With this given RRule  FREQ=DAILY;BYHOUR=13;BYMINUTE=0;BYSECOND=0;DTSTART=20210101;UNTIL=20210105

The  RecurrenceRuleIterator#nextDateTime() returns on the Date without the time component  i.e 20210103,20210104. Would it be possible to get an output of 20210103T130000, 20210104T130000 without changing both DTSTART and UNTIL from All Day to  Floating ?

**Comment by chungonn** [comment]
It seems RecurrenceRuleIterator#nextMillis does provide what I need i.e. Date and Time components. And RecurrenceRuleIterator#nextDateTime() will return the date and time components only if the DTSTART and UNTIL are both are floating DateTime as well. Is this the library design?

**Comment by chungonn** [comment]
Using with the above RRule with an RecurrenceSet. And RecurrenceSet#next only gives date and not the time. And there is no equivalent of RecurrenceRuleIterator#nextMillis in RecurrenceSet.

From tracing the code, the EffectiveInstancesIterator internal instance does provide the time component. Would it be possible to provide the equivalent of RecurrenceRuleIterator#nextMillis in RecurrenceSet? Thanks in advance

**Comment by dmfs** [comment]
That's an interesting use case.
You're right. At present this is not supported. The type of the ` UNTIL`  date must match the type of the start date. However, it seems sensible to support all-day `UNTIL`  dates with any kind of floating start date (all-day or not). I'll consult RFC 5545 to see whether it has to say anything about it.

Regarding the start date format, that one currently determines the format of the resulting dates. Have you considered calling [`DateTime#startOfDay()`](https://github.com/dmfs/rfc5545-datetime/blob/master/src/main/java/org/dmfs/rfc5545/DateTime.java#L743) on the start date before creating the iterator? Once the library supports all-day `UNTIL` with data-&-time start dates that will be sufficient.

It might be quite unexpected when the iterator starts returning date and time when passing and all-day date.

**Comment by dmfs** [comment]
[RFC 5545, Page41](https://www.rfc-editor.org/rfc/rfc5545) states:

> The value of the UNTIL rule part MUST have the same value type as the "DTSTART" property.

"value type" refers to the `VALUE`  value, i.e. ` DATE`  or `DATE-TIME`, so the strict interpretation of the lib seems reasonable. However, I guess in lax modes we could be relaxed about that and accept date-time start values when the `UNTIL` value is "all-day"/`DATE`. 

That, however requires some changes in the logic.

You see, `UNTIL` is interpreted as "inclusive". That means, any instance that matches the `UNTIL` date is still part of the result. What happens when the rule generates `20230305T120000` and `UNTIL=20230305` ? With the naive approach the instance would not be emitted because `UNTIL=20230305` would be interpreted as `UNTIL=20230305T000000` and the instance would be after that. So we need to convert `UNTIL=20230305` to `UNTIL=20230305T235959` in that case to match the expectation.

**Comment by chungonn** [comment]
Hi Marten,

Thanks for the clarification and the explanation and sorry for troubling
you to dig into the RFC 5545 specs as well.

Firstly, thanks for the lib-recur library and I do find the 0.13.0 APIs are
more intuitive. 👏👍

Currently, I am updating my Scala library which uses  lib-recur from 0.12.2
to 0.13.0 and migrate away from the deprecated classes to the new classes
in the *iterable* package. Also in my library with lib-recur 0.12.2
version, I did programmatically add the start--of-day time to the StartDate
when creating a *RecurrenceList* instance.

When you highlight that *UNTIL is interpreted as "inclusive"*, it all makes
sense. If you were to do the expansion -* UNTIL=20230305 to *
*UNTIL=20230305T235959*, it would cause you further trouble to handle
day-light saving issues.

What the library has is good enough for me. I can workaround with the
nextMillis to get the Date and Time component which will meet my needs. As
I mentioned in the post earlier, it would help if there is a
*RecurrenceSet#nextMillis
*an equivalent to *RecurrenceRuleIterator#nextMillis* which returns both
date and time components.

Before I posted the issue, I did use https://jakubroztocil.github.io/rrule/
using *RRule String* tab, it did return both Date + Time components even
when the UNTIL is a date.  I suppose that the library is only RFC 2445
compliant.

To sum it up, I don't think you need to change any of your work or logic. I
would like to request for a new api *RecurrenceSet#nextMillis *which
returns both date + time component :)

Since my existing library is using the 0.12.2 version, and I am just
future-proofing my library with the changes in lib-recur, which means I
have the luxury of time to wait for the requested API.

Thanks once again for looking into my issue and providing the
clarifications and explanations.

**Comment by dmfs** [comment]
It's actually a bit easier than I thought. Instead of converting `UNTIL=20230305` to `UNTIL=20230305T235959`, I can just set hour, minute and second of the generated instance to `0` when comparing the values (i.e. get the start of the day). I don't need to worry about day-light saving, because rule expansion is always performed in local time.

I've created PR #117. I'll sleep over it before merging. Right now I can't see any problem with it.

**Comment by chungonn** [comment]
Big Thanks!  🙏

**Comment by dmfs** [comment]
published in 0.15.0

**Comment by chungonn** [comment]
Hi Marten,

I have tested and it works! Thanks once again :)
